pkg/controllers/management/rkeworkerupgrader/upgrade.go

```diff
@@ -374,6 +393,79 @@ func (uh *upgradeHandler) toUpgradeCluster(cluster *v3.Cluster) (bool, error) {
 	return false, nil
 }
 
+func (uh *upgradeHandler) restore(cluster *v3.Cluster) error {
+	nodes, err := uh.nodeLister.List(cluster.Name, labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	// allowed to use goroutine?
+	var errgrp errgroup.Group
+	nodesQueue := util.GetObjectQueue(nodes)
+	for w := 0; w < 5; w++ {
+		errgrp.Go(func() error {
+			var errList []error
+			for node := range nodesQueue {
+				node := node.(*v3.Node)
+				if node.Status.NodeConfig != nil && workerOnly(node.Status.NodeConfig.Role) {
+					if node.Status.NodePlan.Version != cluster.Status.NodeVersion {
+						if err := uh.processNode(node, cluster, false); err != nil {
+							errList = append(errList, err)
+						}
+						logrus.Infof("cluster [%s]: updated node [%s] for restore, plan version [%v]", cluster.Name,
+							node.Name, cluster.Status.NodeVersion)
+					}
+				}
+			}
+			return util.ErrList(errList)
+		})
+	}
+	if err := errgrp.Wait(); err != nil {
+		return err
+	}
+
+	clusterCopy := cluster.DeepCopy()
+	clusterCopy.Annotations[clusterprovisioner.RkeRestoreAnnotation] = "false"
+	if _, err := uh.clusters.Update(clusterCopy); err != nil {
+		if !errors.IsConflict(err) {
+			return err
+		}
+		return uh.retryClusterUpdate(cluster.Name)
+	}
+	return nil
+}
+
+func (uh *upgradeHandler) retryClusterUpdate(name string) error {
+	backoff := wait.Backoff{
+		Duration: 100 * time.Millisecond,
+		Factor:   1,
+		Jitter:   0,
+		Steps:    7,
+	}
+
+	return wait.ExponentialBackoff(backoff, func() (bool, error) {
+		cluster, err := uh.clusters.Get(name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		if cluster.Annotations[clusterprovisioner.RkeRestoreAnnotation] != "false" {
+			clusterCopy := cluster.DeepCopy()
+			clusterCopy.Annotations[clusterprovisioner.RkeRestoreAnnotation] = "false"
+
+			_, err = uh.clusters.Update(clusterCopy)
+			if err != nil {
+				logrus.Debugf("cluster [%s] restore: error resetting restore annotation %v", cluster.Name, err)
+				if errors.IsConflict(err) {
+					return false, nil
+				}
+				return false, err
+			}
+		}
+		return true, nil
+	})
+
+}
```

Comment:
This is going to try multiple times and it replicates what the norman framework would already do. Why do we need another backoff beyond what the norman framework provides?

Reply:
It's only going to retry only if the retries aren't sufficient from norman. Despite these retries being present in our framework, there are multiple times I've seen these conflict errors in logs

Reply:
I think this is fine, if the cluster comes back through it's going to reprocess all nodes and on a large cluster will eat a large chunk of time and CPU.